src/Components/Web/src/WebRenderer.cs

```diff
@@ -34,25 +34,50 @@ public abstract class WebRenderer : Renderer
         JSComponentInterop jsComponentInterop)
         : base(serviceProvider, loggerFactory)
     {
-        _serviceProvider = serviceProvider;
         _interopMethodsReference = DotNetObjectReference.Create(
             new WebRendererInteropMethods(this, jsonOptions, jsComponentInterop));
 
         // Supply a DotNetObjectReference to JS that it can use to call us back for events etc.
         jsComponentInterop.AttachToRenderer(this);
-        var jsRuntime = _serviceProvider.GetRequiredService<IJSRuntime>();
-        jsRuntime.InvokeVoidAsync(
+        var jsRuntime = serviceProvider.GetRequiredService<IJSRuntime>();
+        _attachTask = jsRuntime.InvokeAsync<int>(
             "Blazor._internal.attachWebRendererInterop",
-            RendererId,
             _interopMethodsReference,
             jsComponentInterop.Configuration.JSComponentParametersByIdentifier,
-            jsComponentInterop.Configuration.JSComponentIdentifiersByInitializer).Preserve();
+            jsComponentInterop.Configuration.JSComponentIdentifiersByInitializer)
+            .AsTask();
     }
 
     /// <summary>
     /// Gets the identifier for the renderer.
     /// </summary>
-    protected int RendererId { get; init; } // Only used on WebAssembly. Will be zero in other cases.
+    /// <remarks>
+    /// Accessing <see cref="RendererId"/> before the renderer is attached will throw an <see cref="InvalidOperationException"/>. Call
+    /// <see cref="WaitUntilAttachedAsync"/> to wait until the renderer gets attached to the browser.
+    /// </remarks>
+    protected int RendererId
+    {
+        get
+        {
+            if (!_attachTask.IsCompletedSuccessfully)
+            {
+                throw new InvalidOperationException($"'{nameof(RendererId)}' does not have a value until {nameof(WaitUntilAttachedAsync)} completes successfully.");
+            }
+
+            return _attachTask.Result;
+        }
+
+        [Obsolete($"The renderer ID gets assigned automatically upon construction.")]
+        init { /* No-op */ }
+    }
+
+    /// <summary>
+    /// Waits until the renderer is attached to the browser. The renderer must be attached before
+    /// renders can be processed.
+    /// </summary>
+    /// <returns></returns>
+    public Task WaitUntilAttachedAsync()
+        => _attachTask;
```

Comment:
My other comment explains the context behind this change. I've made it so that the `RendererId` is allocated on the client side.

Fun fact: Before this PR, the mechanism assigning `RendererId`s was broken. The JS interop call passing `RendererId` to the browser was initiated in the constructor, which runs before the derived constructors run. So, no matter what value was assigned to `RendererId` in derived classes, the `RendererId` passed to JS was always 0.